src/navigation_plugin.cpp

```diff
@@ -68,84 +64,73 @@ void constructConstraint(std::vector<geo::Vector3>& chull, std::stringstream& co
 
 // ----------------------------------------------------------------------------------------------------
 
+
 //
 /**
  * @brief constructShapeConstraint
- * @param r r must be currently reading an array of shapes
+ * @param shape shape of the volume in entity frame
  * @param entity_pose pose of the entity
  * @return constraint string
  */
-std::string constructShapeConstraint(tue::config::Reader& r, const geo::Pose3D& entity_pose)
+std::string constructShapeConstraint(geo::ShapeConstPtr& shape, const geo::Pose3D& entity_pose)
 {
     std::stringstream shape_constraint;
 
-    bool first_sub_shape = true;
-    while(r.nextArrayItem())
-    {
-        std::stringstream sub_shape_constraint;
+    geo::Shape shape_tr;
+    shape_tr.setMesh(shape->getMesh().getTransformed(entity_pose)); // get shape in map frame.
+    std::vector<geo::Vector3> points = shape_tr.getMesh().getPoints();
 
-        if (r.readGroup("box", tue::config::OPTIONAL))
-        {
-            std::vector<geo::Vector3> chull; // In MAP frame
+    // Calculate cluster convex hull
+    float z_min = 1e9;
+    float z_max = -1e9;
 
-            geo::Vector3 min, max;
+    // Calculate z_min and z_max of cluster
+    std::vector<geo::Vec2f> points_2d;
+    for(std::vector<geo::Vector3>::const_iterator it = points.begin(); it != points.end(); ++it)
+    {
+        points_2d.push_back(geo::Vec2f(it->x, it->y));
 
-            if (r.readGroup("min"))
-            {
-                r.value("x", min.x);
-                r.value("y", min.y);
-                r.value("z", min.z);
-                r.endGroup();
-            }
+        z_min = std::min<float>(z_min, it->z);
+        z_max = std::max<float>(z_max, it->z);
+    }
 
-            if (r.readGroup("max"))
-            {
-                r.value("x", max.x);
-                r.value("y", max.y);
-                r.value("z", max.z);
-                r.endGroup();
-            }
+    ed::ConvexHull chull;
+    ed::convex_hull::createAbsolute(points_2d, z_min, z_max, chull);
 
-            chull.push_back(min);
-            chull.push_back(geo::Vector3(max.x, min.y, 0));
-            chull.push_back(max);
-            chull.push_back(geo::Vector3(min.x, max.y, 0));
+    constructConstraint(chull, shape_constraint);
+    return shape_constraint.str();
+}
 
-            // Transform to map frame, and add to entity constraint
-            transformChull(entity_pose, chull);
-            constructConstraint(chull, sub_shape_constraint);
+// ----------------------------------------------------------------------------------------------------
 
-            r.endGroup();
-        }
-        else if (r.readGroup("polygon", tue::config::OPTIONAL))
-        {
-            std::vector<geo::Vector3> chull; // In MAP frame
+//
```

Comment:
Either needs an empty line below it, or (imho preferably) kill lines 104 and 105, as the docstring below already serves as a separator.